### PR TITLE
responsive layout

### LINKED ISSUE #22: Create responsive design
The UI looks really nice on a desktop pc. I tried it on my phone and some tablets but it looks bad there. I want to use the webinterface with a tablet in the real fotobox.

**Design Problems**:
- "Creating foto"-button in the middle of the text on the front page
-  Width / Height of the hole site is wrong (I can scroll left / right or up / down on my mobile devices)
-  Gallery to small or rather only on a half screen
-  QR Code to big on mobile devices in landscape mode
-  Various bugs after rotating the phone / tablet

### PR DESCRIPTION
**Pull request by sualko**
This is work in progress, but the startscreen and the gallery is already responsive.

fix #22 

2019-09-24 19 50 12 localhost 18aa8cbc9815
2019-09-24 19 50 33 localhost 6495004e36c1

### CODE REVIEW

**Comment by andi34** [comment]
Great! Thanks a lot for all your work. Going to review your open pull request tomorrow morning.

**Comment by sualko** [comment]
Do you have an idea why the sidenav style is missing in the sass file?

**Comment by andi34** [comment]
Seems like I forgot to add it there, I always only used the style.css file

**Comment by andi34** [comment]
I am still learning all the stuff here. Learning by doing to be true 😅🤗

**Comment by tobiashaas** [comment]
Ich bin auch gerade dabei dass alles Responsive umzusetzen! - leider zu langsam gewesen :) 
Werde aber ähnlich wie das BlueGrey Theme, einen anderen style bauen ;-) 
Könnte man ja dann mit dazu nehmen! 

Kann mir noch einer erklären für was genau die Sass Dateien sind? Hab das noch nicht ganz verstanden ... 

Liebe Grüße

**Comment by andi34** [comment]
@tobiashaas klar wegen dem Farbtheme. Config kann man entsprechend dann umbauen das man ein Auswahlmenü hat. kann ich übernehmen den Part sobald dein Theme fertig ist.

Bezüglich sass muss ich leider auch passen.

**Comment by sualko** [comment]
Sass is a superset of css. It adds enclosures, variables, functions, includes and a lot more to css. On the downside it requires a transformation step, because browsers don't support sass. Therefore you should only edit those sass files and never the produced css files.

If you plan to change the style, please wait until this pr is finished, because in order to make the whole thing responsive I have to make some structural changes.

**Comment by andi34** [comment]
Thanks a lot for the explanation and your work!

**Comment by sualko** [comment]
2019-09-25 14 04 06 localhost 572a8b35771a
2019-09-25 14 03 15 localhost 2d5debbe36fb
2019-09-25 14 03 49 localhost 705a0fa575d1
2019-09-25 14 03 37 localhost c470da7482c3

This pull request needs some really good tests, because I'm not sure if I missed something. Also I have to fix the bluegray theme.

After that I would start to restructure the code to have a foundation for some enhancements regarding the collage.

**Comment by andi34** [comment]
Could the sidenavbar (filter menu) be a little bit transparent?

**Comment by andi34** [comment]
Also as mentioned I guess the video code is still needed, e.g. to use picamera for preview

**Comment by tobiashaas** [comment]
Ich finde es gut dass die Sidebar nicht transparent ist. 
Hatte letztens einen Hintergrund bei dem man kaum lesen konnte wie der Filter heißt 🙈

**Comment by andi34** [comment]
currently i only get a white page

Screenshot_2019-09-25_18-36-57

**Comment by sualko** [comment]
Did you compile the sass files? If not, move to the resources directory and run `npm build` or `yarn build`.

**Comment by andi34** [comment]
got it working now:
```
npm install -g sass
sass resources/sass/style.scss resources/css/style.css 
```

What i noticed  so far:
- the print confirmation dialogue is missing now
- countdown animation needs adjustment (was smoother before)
- QR-Code is missing the text on desktop view, on mobile view still oversized
- Home-Button and Fork-Me logo conflict on mobile view

Besides the nitpicks it looks great :) Thanks a lot!

**Comment by sualko** [comment]
I think I'm done so far, but please give it a really good test, because I don't know if I missed a feature or something.

**Comment by andi34** [comment]
https://github.com/andreknieriem/photobooth/commit/ab2ebf3f551d194187b9918a926ad75aa70468fd#diff-7b96a31b94239de857d8034ed71b696eR834

- css parts for print confirmation missing.
- chromakeying.php and resources/css/chromakeying.css  might need adjustments too (at least for theming?)
- admin panel also need adjustments for theming?

**Comment by sualko** [comment]
Please give it another try. Thanks for testing btw.

**Comment by andi34** [comment]
only getting a white page, css files recreated as on https://github.com/sualko/photobooth/pull/1

**Comment by andi34** [comment]
ok, must be something on my end (maybe the theme change commit). on your branch now with recreated css files, but nothing apears:

Screenshot_2019-09-26_14-57-20

**Comment by sualko** [comment]
Did you do a force reload? Otherwise please check the js console. I see no possibility to debug this issue on my side.

**Comment by andi34** [comment]
Uncaught SyntaxError: Unexpected token 'default'
theme.js:1 Uncaught ReferenceError: theme is not defined
    at theme.js:1

**Comment by sualko** [comment]
I guess this is related to your own changes in your pr.

**Comment by andi34** [comment]
Fixed now.

**Comment by andi34** [comment]
Cosmetical thing regarding QR:

right after picture was taken
Screenshot_2019-09-26_15-35-26

vs

qr from gallery
Screenshot_2019-09-26_15-34-46

From Gallery the text is missing

**Comment by sualko** [comment]
I know that the text is missing, but this was also the case before.

**Comment by andi34** [comment]
Ok, thought it'll be identical after the rework.

So far so good. Haven't found any issues yet. Going to test it further tomorrow.

Thanks a lot again!

It'll be much work for you to integrate a button to download the picture directly?
Would be gread to have i as an option instead using the QR, i've tried in the past but failed completely :see_no_evil:

**Comment by sualko** [comment]
> Ok, thought it'll be identical after the rework.

We put it in another pr :wink:

> It'll be much work for you to integrate a button to download the picture directly?

What do you mean? Where should it be downloaded? A photobooth is meant to be autark and all images will be inside the `images/` folder. Can you elaborate a bit more on that?

**Comment by andi34** [comment]
> > Ok, thought it'll be identical after the rework.
> 
> We put it in another pr wink

:+1: 

> 
> > It'll be much work for you to integrate a button to download the picture directly?
> 
> What do you mean? Where should it be downloaded? A photobooth is meant to be autark and all images will be inside the `images/` folder. Can you elaborate a bit more on that?

The QR-Code redirects people to download the picture from the Photobooth.
But if a user has access to the gallery from his/her phone, it'll be better imo to have a download button on the open picture to directly download it.

**Comment by andi34** [comment]
Some screenshots from OnePlus 6:
Screenshot_20190926-175042
Screenshot_20190926-175101
Screenshot_20190926-175048


Odd thing: desktop view start page
Screenshot_20190926-175753


Odd thing mobile view:
Keying not possible because it's looking for the thumb image... thumb/ don't need to be inside the URL, not sure where it comes from. Works fine on desktop page.
Screenshot_20190926-180033
Screenshot_20190926-180052

**Comment by andi34** [comment]
The issue with keying on mobile view is solved by https://github.com/andi34/photobooth/commit/d6e5e77f2ba35d85a1656cc231d4488d707abc25

**Comment by sualko** [comment]
> Ok, thought it'll be identical after the rework.

My plan is to replace the php side generated qr code by a javascript solution and during this work I would also refactor the qr code dialog.

> But if a user has access to the gallery from his/her phone, it'll be better imo to have a download button on the open picture to directly download it.

I think this should go to the standalone gallery. But first we have to merge this pr and restructure the repository.

> Odd thing "desktop site" in chrome settings:
on the start page the filter menu is always visible

That's weird, but I think this is not a normal use case. We should keep that in mind for a time if we have nothing else to do :smile:.

> Keying not possible because it's looking for the thumb image

I will cherry-pick your proposed solution in a minute.

**Comment by andi34** [comment]
@sualko https://github.com/andi34/photobooth/commits/triggerbutton

This branch includes everything without personal stuff which will conflict.
Todo on this:
- some translation optimization on 2 commits
- adjust the cups button?
- adjust the css changes to be in the sass file

**Comment by andi34** [comment]
#### Test this changes (and everything else currently open for review)
Make sure your local clone does ignore filemode changes:
```
cd /var/www/html
git config core.fileMode false
sudo git config core.fileMode false
```
Now add my Github copy as remote and fetch the source:
```
git remote add andi34 https://github.com/andi34/photobooth
git fetch andi34
```
Make sure you have no local changes which will be lost, if you have you need to commit them. Best ist to keep them in a seperate branch you can go back to later:
```
git add --all
git commit -a -m "All my local changes"
git checkout -b myBackupBranch
```
Now switch to my source:
```
git checkout andi34/triggerbutton
```

You might need to set the permissions again on ```/var/www``` as mentioned in the README.
Important: Make sure deleting ```admin/config.json``` if exist!

In case you need to switch back to your local source you only need to checkout to your backup branch which we've created before.
```
git checkout myBackupBranch
```

Questions? Let me know!

**Comment by sualko** [comment]
How are your changes related to this pr? One pr should have one feature or fix. Otherwise it will be a nightmare to keep the overview and review the code. Can we merge this feature and look at your stuff afterwards?

**Comment by andi34** [comment]
> How are your changes related to this pr? One pr should have one feature or fix. Otherwise it will be a nightmare to keep the overview and review the code. Can we merge this feature and look at your stuff afterwards?

Sure we can, thought it doesn't hurt if the other stuff is looked at too while people test anyway.
Also i though "Add trigger button function" and "Add option for CUPS button" need to be adjusted too for responsive layout.

To let people only test this PR you'd like to add the recreated css files? Guess it'll be problematic for some to generate them.
And maybe also add https://github.com/andi34/photobooth/commit/c86d129e9449924630cd640eae44fdb2c875db55 as it's part of the responsive theme change (was missing here https://github.com/andreknieriem/photobooth/pull/78/commits/dd1d35dfef7a66eb166dfaf08f1082b9c7ff8ae5 )?

**Comment by sualko** [comment]
> Sure we can, thought it doesn't hurt if the other stuff is looked at too while people test anyway.

I think we have to distinguish between developer and tester. Pull requests should be tested by developers, because you not only test functionality, but also the quality of the code. Testers are testing pre-releases, like a beta version before the stable release is released. Therefore it's also bad practice to include build files into git, because it clutters the diff. If someone likes to test this pr, please use the following style files: [css.zip](https://github.com/andreknieriem/photobooth/files/3662196/css.zip).

Would really like to get everything done one after the other, because e.g. I have some questions to the new `use_pi_cam` config.

**Comment by sualko** [comment]
If you look at my [refactoring branch](https://github.com/sualko/photobooth/tree/feature/refactor) you know why it doesn't make sense to include build files. This way it's easier for the developer to keep e.g. all dependencies at the latest release.

**Comment by andi34** [comment]
> > Sure we can, thought it doesn't hurt if the other stuff is looked at too while people test anyway.
> 
> I think we have to distinguish between developer and tester. Pull requests should be tested by developers, because you not only test functionality, but also the quality of the code. Testers are testing pre-releases, like a beta version before the stable release is released. Therefore it's also bad practice to include build files into git, because it clutters the diff. If someone likes to test this pr, please use the following style files: [css.zip](https://github.com/andreknieriem/photobooth/files/3662196/css.zip).
> 

Ok, no problem.
From what i've tested i didn't found any issues. Codewise looks a lot cleaner now.
One question would be to move the styling from chromakeying.php (e.g. ```<div style="width:600px; max-width: 100%; margin:0 auto;background-color:#000000;border:4px solid black;">
``` ) also to the sass files ?

Browser console tells:
```The jquery.gsap.js plugin requires the TweenMax (or at least TweenLite and CSSPlugin) JavaScript file(s).``` Not sure if TweenLite is still needed, but removed here.

> Would really like to get everything done one after the other, because e.g. I have some questions to the new `use_pi_cam` config.

Was a request in the comments [here](https://www.andrerinas.de/tutorials/raspberry-pi-einen-dslr-weddingphotobooth-erstellen.html) to have pi cam settings as inline notes, but thought it'll be better to make this switchable via config. Also could be used later for other stuff too if needed.

**Comment by sualko** [comment]
Tween lite is now back. Would love to reduce the number of dependencies, but I think here I was a bit fast.

**Comment by andi34** [comment]
> Tween lite is now back. Would love to reduce the number of dependencies, but I think here I was a bit fast.

Thanks :)

I've send you a mail with some information.

**Comment by andi34** [comment]
Found one issue: show date in caption doesn't work.
We need to remove 
```
.pswp__caption {
  display: none; }
```

File: index.php [hunk 1]
```diff
@@ -75,6 +75,10 @@
 				<!-- accesskey to take a photo using alt+p (or use an external button)? -->
 				<a href="#" <?php if($config['use_gpio_button']){ ?>accesskey="p"<?php } ?> class="btn takePic"><i class="fa fa-camera"></i> <span data-l10n="takePhoto"></span></a>
 			</div>
+
+			<?php if($config['show_fork']): ?>
+			<a target="_blank" href="https://github.com/andreknieriem/photobooth"><img style="position: absolute; top: 0; right: 0; border: 0;" src="https://camo.githubusercontent.com/652c5b9acfaddf3a9c326fa6bde407b87f7be0f4/68747470733a2f2f73332e616d617a6f6e6177732e636f6d2f6769746875622f726962626f6e732f666f726b6d655f72696768745f6f72616e67655f6666373630302e706e67" alt="Fork me on GitHub" data-canonical-src="https://s3.amazonaws.com/github/ribbons/forkme_right_orange_ff7600.png"></a>
```
Comment (andi34) [inline index.php]: https://github.com/andi34/github-ribbons-css

Switch to a css ribbon?

Comment (sualko) [reply index.php]: Put it on our todo list :joy:

Comment (andi34) [reply index.php]: https://github.com/andreknieriem/photobooth/issues/68 Added to WIP

**Comment by andi34** [comment]
IMO ready to merge. It'll be ok?

**Comment by sualko** [comment]
> It'll be ok?

I hope so :-)

**Comment by sualko** [comment]
Did you rebase this pr? This messed everything up with my refactoring branch...

**Comment by sualko** [comment]
I think you cherry-picked the whole branch :cry: Please never ever do this again. I have now no idea how I should integrate all my changes from the refactor branch.

**Comment by andi34** [comment]
Rebased and merged. Makes it easier to go back to a specific commit if something breaks.
Don't worry, I know git really good.
@sualko use https://github.com/andi34/photobooth/tree/feature/refactor

**Comment by sualko** [comment]
> Don't worry, I know git really good.

I don't think so. There is absolutely no reason for such an action.

> Makes it easier to go back to a specific commit if something breaks

Even after a merge you can go back to every commit you like and in this case even a fast-forward had been possible. Now if you compare different branches or forks, git shows you all commits twice.

**Comment by andi34** [comment]
From my work on android i know there'll be more trouble compared merge with commits rebased. Also all the "merge xyz" commits which get created imo don't look nice in history.
On rebase and merge the risk of a mismerge is a lot lower.

But i see it's solved now. Sorry again.

**Comment by sualko** [comment]
> Rebased and merged

If you know git that well, than you know that this is not possible.

> all the "merge xyz" commits which get created imo don't look nice in history.

fast-forward? Who is interested in a nice history? Nobody (or use gitk or something else). It's more important that I know that my changes are in a branch. If you rebase something I have to check if you changed my commits, because they will get a new id.

> From my work on android

Please look at other open source project and see how they handle merges and thing about why.

**Comment by andi34** [comment]
@sualko you're familiar with Gerrit?
I could talk to my friend if it's ok to hijack https://gerrit.unlegacy-android.org/ for the refactoring. This would also make it easier for you to adjust changes if needed.

**Comment by sualko** [comment]
I don't know Gerrit, but it was developed for svn and therefore I don't know how helpful this tool is. I can only tell you that I contributed in a lot of open source projects (even large ones like Nextcloud) and I'm the main developer of JSXC which has several thousands of active users.

I understand that you like to have a linear commit history, but I think the disadvantages are to huge (no signing, no comparison, difficult merges, ...). Please look at a fast-forward merge if you like to keep the history flat.

The git branch overview also doesn't look that good:
2019-09-27 15 40 21 github com 98c0b29e17b5

**Comment by andi34** [comment]
Gerrit is a really useful reviewing tool, especially as long as changes are WIP. You can update changes until they are ready, always return to a previous patchset of the change.
A review system allows other devs to vote on made changes and discuss changes if needed.
To submit changes to Gerrit every commit needs a unique Commit-ID which will be added at git commit if it's setup. If a change is cherry-picked to a different project it'll remain the unique Commit ID.
From Gerrit changes can be submitted to e.g. GitHub.

But well, guess we'll forget about that here an concentrate on the important things.
On your next PR I can click on "merge" instead "rebase and merge".

**Comment by andi34** [comment]
And i am sorry GitHub doesn't show the merge in the network graph. But on top of the PR you see that everything is in.
Screenshot_20190927-162123__01

**Comment by thymon13** [comment]
Works with Google chrome.

But I use "FullScreenWebBrowser" app to my Android Tablet.
the display has a problem with that : https://youtu.be/PSsM1K6z3hs

**Comment by andi34** [comment]
You could try opera, from the screenshots on play store it has an option for fullscreen.
Screenshot_20190927-210814

**Comment by andi34** [comment]
Another solution might https://play.google.com/store/apps/details?id=de.ozerov.fully

**Comment by thymon13** [comment]
@andi34 Opera is not real full screen. I tested
https://play.google.com/store/apps/details?id=de.ozerov.fully --> Same problem

**Comment by andi34** [comment]
Which webview used?
Screenshot_20190927-213531
Screenshot_20190927-213522

**Comment by thymon13** [comment]
@andi34 I do not have "webView implementation" in Developer options
😢

**Comment by thymon13** [comment]
@andi34 You are the BEST !! Thanks a lot !!

...I'm really not used to android

**Comment by andi34** [comment]
Changing it fixed it?

**Comment by thymon13** [comment]
YEEES !! I have updated "webview android system" from PlayStore before.

**Comment by andi34** [comment]
Good to.know, thanks for the feedback 🙂

**Comment by thymon13** [comment]
I'have found a bug.

I take a picture -> I click on QR code -> I click on screen to back --> I click home : OK
I take a picture -> I click on QR code --> I wait a few minute -> I click on screen to back > I click home : NOK I can't back to Home.

**Pull request merged**